src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs

```diff
@@ -242,123 +241,97 @@ private unsafe void ReadDirectoryChangesCallback(uint errorCode, uint numBytes,
             }
         }
 
-        private void ParseEventBufferAndNotifyForEach(byte[] buffer)
+        private unsafe void ParseEventBufferAndNotifyForEach(byte[] buffer)
         {
             Debug.Assert(buffer != null);
             Debug.Assert(buffer.Length > 0);
 
-            // Parse each event from the buffer and notify appropriate delegates
-
-            /******
-                Format for the buffer is the following C struct:
-
-                typedef struct _FILE_NOTIFY_INFORMATION {
-                   DWORD NextEntryOffset;
-                   DWORD Action;
-                   DWORD FileNameLength;
-                   WCHAR FileName[1];
-                } FILE_NOTIFY_INFORMATION;
+            ReadOnlySpan<char> oldName = ReadOnlySpan<char>.Empty;
 
-                NOTE1: FileNameLength is length in bytes.
-                NOTE2: The Filename is a Unicode string that's NOT NULL terminated.
-                NOTE3: A NextEntryOffset of zero means that it's the last entry
-            *******/
-
-            // Parse the file notify buffer:
-            int offset = 0;
-            int nextOffset, action, nameLength;
-            string oldName = null;
-            string name = null;
+            // Note that the buffer is pinned for the lifetime of the watcher.
+            // We have to pin again here to get the pointer.
+            Interop.Kernel32.FILE_NOTIFY_INFORMATION* info;
+            fixed (byte* b = buffer)
+            {
+                info = (Interop.Kernel32.FILE_NOTIFY_INFORMATION*)b;
```

Comment:
Leaking out the pinned pointer outside the scope looks dangerous. I know that you have comment that this is safe because of the buffer is pinned elsewhere ... but why not just pinned for the whole duration to make it look obviously right? I won't cost you anything real.

Reply:
I assume you mean for the duration of this method. I can do that. I didn't originally to avoid another level of nesting.